crates/stackable-webhook/src/webhooks/conversion_webhook.rs

```diff
@@ -0,0 +1,201 @@
+use std::fmt::Debug;
+
+use async_trait::async_trait;
+use axum::{Json, Router, routing::post};
+use k8s_openapi::{
+    ByteString,
+    apiextensions_apiserver::pkg::apis::apiextensions::v1::{
+        CustomResourceConversion, CustomResourceDefinition, ServiceReference, WebhookClientConfig,
+        WebhookConversion,
+    },
+};
+// Re-export this type because users of the conversion webhook server require
+// this type to write the handler function. Instead of importing this type from
+// kube directly, consumers can use this type instead. This also eliminates
+// keeping the kube dependency version in sync between here and the operator.
+pub use kube::core::conversion::ConversionReview;
+use kube::{
+    Api, Client, ResourceExt,
+    api::{Patch, PatchParams},
+};
+use snafu::{ResultExt, Snafu, ensure};
+use tokio::sync::oneshot;
+use tracing::instrument;
+use x509_cert::Certificate;
+
+use super::{Webhook, WebhookError};
+use crate::WebhookServerOptions;
+
+#[derive(Debug, Snafu)]
+pub enum ConversionWebhookError {
+    #[snafu(display("failed to send initial CRD reconcile heartbeat"))]
+    SendInitialReconcileHeartbeat,
+
+    #[snafu(display("failed to patch CRD {crd_name:?}"))]
+    PatchCrd {
+        source: kube::Error,
+        crd_name: String,
+    },
+}
+
+pub struct ConversionWebhook<H> {
+    crds_and_handlers: Vec<(CustomResourceDefinition, H)>,
+    disable_crd_maintenance: bool,
+    client: Client,
+    /// The field manager used when maintaining the CRDs.
+    field_manager: String,
+    // This channel can only be used exactly once. The sender's send method consumes self, and
+    // as such, the sender is wrapped in an Option to be able to call take to consume the inner
+    // value.
+    initial_reconcile_tx: Option<oneshot::Sender<()>>,
+}
+
+impl<H> ConversionWebhook<H> {
+    pub fn new(
+        crds_and_handlers: impl IntoIterator<Item = (CustomResourceDefinition, H)>,
+        disable_crd_maintenance: bool,
+        client: Client,
+        field_manager: String,
+    ) -> (Self, oneshot::Receiver<()>) {
+        let (initial_reconcile_tx, initial_reconcile_rx) = oneshot::channel();
+
+        (
+            Self {
+                crds_and_handlers: crds_and_handlers.into_iter().collect(),
+                disable_crd_maintenance,
+                client,
+                field_manager,
+                initial_reconcile_tx: Some(initial_reconcile_tx),
+            },
+            initial_reconcile_rx,
+        )
```

Comment:
note: This looks fairly cluttered, can we make this nicer to look at.

Reply:
I moved the Self{} into it's own variable in 9494797, hope that works